vm/src/bench/kernel.rs

```diff
@@ -0,0 +1,421 @@
+        Ok(vec![true; vis.len()])
+    }
+
+    // NOT forwarded
+    fn verify_seal(&self, vi: &SealVerifyInfo) -> Result<bool> {
+        let charge = self.test_config.price_list.on_verify_seal(vi);
+        let _ = self.inner_kernel.charge_gas(&charge.name, charge.total())?;
+        Ok(true)
+    }
+
+    // NOT forwarded
+    fn verify_post(&self, vi: &WindowPoStVerifyInfo) -> Result<bool> {
+        let charge = self.test_config.price_list.on_verify_post(vi);
+        let _ = self.inner_kernel.charge_gas(&charge.name, charge.total())?;
+        Ok(true)
+    }
+
+    // NOT forwarded
+    fn verify_consensus_fault(
+        &self,
+        h1: &[u8],
+        h2: &[u8],
+        extra: &[u8],
+    ) -> Result<Option<ConsensusFault>> {
+        let charge =
+            self.test_config.price_list.on_verify_consensus_fault(h1.len(), h2.len(), extra.len());
+        let _ = self.inner_kernel.charge_gas(&charge.name, charge.total())?;
+        Ok(None)
+    }
+
+    // NOT forwarded
+    fn verify_aggregate_seals(&self, agg: &AggregateSealVerifyProofAndInfos) -> Result<bool> {
+        let charge = self.test_config.price_list.on_verify_aggregate_seals(agg);
+        let _ = self.inner_kernel.charge_gas(&charge.name, charge.total())?;
+        Ok(true)
+    }
+
+    // NOT forwarded
+    fn verify_replica_update(&self, rep: &ReplicaUpdateInfo) -> Result<bool> {
+        let charge = self.test_config.price_list.on_verify_replica_update(rep);
+        let _ = self.inner_kernel.charge_gas(&charge.name, charge.total())?;
+        Ok(true)
+    }
+}
+
+impl<C> DebugOps for BenchKernel<C>
+where
+    C: CallManager,
+{
+    fn log(&self, msg: String) {
+        self.inner_kernel.log(msg)
+    }
+
+    fn debug_enabled(&self) -> bool {
+        self.inner_kernel.debug_enabled()
+    }
+
+    fn store_artifact(&self, name: &str, data: &[u8]) -> Result<()> {
+        self.inner_kernel.store_artifact(name, data)
+    }
+}
+
+impl<C> GasOps for BenchKernel<C>
+where
+    C: CallManager,
+{
+    fn gas_used(&self) -> Gas {
+        self.inner_kernel.gas_used()
+    }
+
+    fn charge_gas(&self, name: &str, compute: Gas) -> Result<GasTimer> {
+        self.inner_kernel.charge_gas(name, compute)
+    }
+
+    fn price_list(&self) -> &PriceList {
+        self.inner_kernel.price_list()
+    }
+
+    fn gas_available(&self) -> Gas {
+        self.inner_kernel.gas_available()
+    }
+}
+
+impl<C> MessageOps for BenchKernel<C>
+where
+    C: CallManager,
+{
+    fn msg_context(&self) -> Result<MessageContext> {
+        self.inner_kernel.msg_context()
+    }
+}
+
+impl<C> NetworkOps for BenchKernel<C>
+where
+    C: CallManager,
+{
+    fn network_context(&self) -> Result<NetworkContext> {
+        self.inner_kernel.network_context()
+    }
+
+    fn tipset_cid(&self, epoch: ChainEpoch) -> Result<Cid> {
+        self.inner_kernel.tipset_cid(epoch)
+    }
+}
+
+impl<C> RandomnessOps for BenchKernel<C>
+where
+    C: CallManager,
+{
+    // NOT forwarded
```

Comment:
I don't think we want to intercept this. It should come from externs which we can control at construction.

Reply:
https://github.com/anorth/fvm-workbench/issues/10